README.md
# FrACT₁₀

[FrACT](https://michaelbach.de/fract/) is a semiautomated vision test battery. FrACT₁₀ is a version that runs in browsers on all plattforms or as webApp. For full details see <https://michaelbach.de/fract/>.

This file is part of FrACT10.
Copyright © 2021 Michael Bach, bach@uni-freiburg.de, <https://michaelbach.de>

## Programming overview
The program is mostly written in Objective-J, a language implemented by <https://www.cappuccino.dev>, and compiled by Cappuccino into JavaScript. [While the Cappuccino site looks a bit lazy, there is a small, select and active community working on it, communicating via Gitter.] Objective-J is modeled on Objective-C and implements much of Apple's GUI tools. Some parts of FrACT₁₀ are also pure JavaScript. Apple's Xcode is used as the development environment. The program is started by opening the project-relative top-level `index.html` in a browser.

For more see [CONTRIBUTING.md](CONTRIBUTING.md).

[![Donate?](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=Z9ERGF97E67AJ)


## Some History
+ 2021-06-01 made open source
+ 2020-05-24 created the github repository
+ 2015 began porting from the "classical" FrACT3.x.


## License
[GNU GPL](LICENSE.md)


History.md
#  History

+ 2023-07-30 implement what2Sweep stuff (contrast or spatial frequency): working, improve GUI, needs better export
+ 2023-07-29 refactor gratings (prepare frequency sweep for BCM)
+ 2023-07-28 fix colorgrating background
+ 2023-07-24 improve more in help panel
+ 2023-07-21 improve help panel
+ 2023-07-06 fix regression in acuity GUI
+ 2023-07-05 BCM: colored grating 1st try
+ 2023-07-04 begin additions for the BCM project
+ 2023-07-03 noise also for contrast, simpler decisions on test category with "isContrast/isAcuity"
+ 2023-07-02 correct a tooltip, improve make script
+ 2023-07-01 begin "embed in noise", Version → 1.0.3
+ 2023-06-27 add preset for color
+ 2023-06-26 add export for color, "gray" → "grey"
+ 2023-06-25 begin color for acuity optotypes
+ 2023-06-24 improve the webApp manifest (e.g. added type in header), no more stray Apple icon, re-check on storing CPColor in userdefaults (missing CPData component), improve make script
+ 2023-05-31 History → separate file
+ 2023-05-29 fix regression: GUI had overlap in an optionally hidden element
+ 2023-05-28 need to increase limits for distance and calibration bar (for demo in lecture hall)
renaming of files (capp→FrACT), now only webApp
+ 2023-05-14 remove setting for window background color (not really needed, crowds) and rearrange some GUI elements
+ 2023-05-05 →webApp only, no more mention of Electron and helpers. Also fix regression with lost text in Acuity-E expl.
+ 2023-05-03 ensure that view witdth/height is correct after window resizing
+ 2023-05-03 refactor: centralise draw touch controls to avoid repeating myself
+ 2023-05-02 add autoRun for optotype & grating contrasts
+ 2023-05-02 simplify TrialHistoryController value entries, improve some names concerned with contrast
+ 2023-04-30 error diffusion for contrast grating optional
+ 2023-04-30 contrast grating looking good
## ↑Gratings

+ 2023-04-22 fix crashing bug with crowding "surounding bars" introduced by "modernising" Javascript
+ 2023-03-26 default gamma value now 2.0, formerly 1.8; fix regression from MiscSpace factorisation; fix inverted ">" sign in vernier overflow
+ 2023-03-25 factorisation: separate MiscLight & MiscSpace from Misc; move "existsUrl" to Misc
+ 2023-03-24 add "Test" to presets to make life easier for me
+ 2023-03-18 slight GUI layout adjustments
+ 2023-03-07 automatic copyright end year from versionstring in About
+ 2023-03-05 more WebApp mods, add field for current value of decimalMarkChar
+ 2023-03-04 no change in the cappuccino code, but modifications for WebApp
## ↑WebApp

+ 2023-02-07 add AutoTest for VA;  gThresholdCorrection4Ascending→global
+ 2023-02-07 correct slope "polarity" for the dispersion estimation, gSlopeCI95→global
+ 2022-11-11 fix missing path in build script call, don't show Terminal – now it really works :)
+ 2022-11-10 make path to build script project-relative
+ 2022-11-08 fix a typo in tooltips, add the `XcodeCapp`-emulating shell script to project (runs via ⌘-B), add `*.xcodeproj` to git
+ 2022-10-28 add field with calculated inch value for distance, slight shifts in GUI
+ 2022-10-27 rename `credit card` to `plastic card`, shift related fields in GUI a little
+ 2022-10-12 add license in About
+ 2022-10-03 "Landolt-C" → "Landolt C"
+ 2022-09-29 fix regression: "var" → "const", should be "let" at two places
+ 2022-09-24 fix rewardPictures: first one offset by half a picture width and always same
+ 2022-09-24 all "var" → "const" (where possible) or at least "let" (modernising JavaScript)
+ 2022-09-18 the new "es2022" compiler allows modern JavaScript, so simplify code with an arrow func
more fun with "es2022", implement some "for x of <array>"
+ 2022-09-11 make Quit/Exit button title depend on operating system (Mac: Quit)
+ 2022-09-02 increase max value of `contrastOptotypeDiameter` from 1500 to 2500
+ 2022-09-01 new compiler allows "let" and "const", begin to use them
## ↑modernised JavaScript

+ 2022-08-30 in "applyESU": set "no clipboard" to avoid nagging question
+ 2022-08-29 correct version from 10.0 to 1.0, improve / add several tooltips, add Electron refs
+ 2022-08-17 increase max value of `contrastOptotypeDiameter` from 500 to 1500
+ 2022-08-11 reword calibration warning, trivial button resizing
+ 2022-08-03 allow larger distance (50 m, for presentations in halls), rename Quit/Exit → Exit/Quit
+ 2022-07-18 make Quit/Exit always available (no longer only in Node), also easier because no nw/electron discriminatin necessary
+ 2022-07-03 select Line-by-line distance between optotypes either ISO or ETDRS
+ 2022-07-02 Line-by-line to allow Landolt Cs in addition to Sloan Letters
+ 2022-06-23 improve wording in About
+ 2022-06-17 improve webview in About (border not faked by textfield etc.),
Help panel also with webview, share buttons between Help and About panels,
fix type inconsistencies in Settings (just for code beauty, had no effect),
simplify "decimalMarkChar" in Settings
+ 2022-06-15 unclutter main screen: move Cappuccino version to About panel, kVersionStringOfFract w/o "Vs"; revamp About a little
+ 2022-06-13 add version of Cappuccino Framework (1.2.2 currently) and display,
fix randomising the pseudorandom sequence
+ 2022-05-28 move Presets pop-up to Settings (where it belongs),
IB "builds for" must be set to "10.12 and later" and NOT to anything later.
+ 2022-05-26 improve "make-XcodeCappSimile.sh", improve Presets, worked on Documentation
+ 2022-05-24 add Presets (ESU and ULV so far), some editing in CONTRIBUTING.md.
BIG addition: new "make-XcodeCapp_and_run.sh" to replace XcodeCapp
## ↑XcodeCapp replaced

+ 2022-05-22 cosmetic change: gaMinimal/Maximal → strokeMinmal/Maximal
gStrokeMinimal / gStrokeMaximal now global, so accessible from dispersion estim.
Add "acuityStartingLogMAR" logic + GUI, good for ultra low vision
+ 2022-05-21 preparation for Presets
+ 2022-04-26 gStrokeMaximal = viewHeight / (5 + 1); // this leaves ½gap margin around optotype

_make-FrACT.sh
#!/bin/zsh

# 2023-06-24 mv rather than rm avoids creation of the *2 files 
# 2023-03-04 added jakefile creation to cooperate with the webApp
# 2021-08-01 moved "rm …capp" before "jake…"
# 2021-02-04 begun

cd ${0:a:h} # go to the starting directory
# pwd

#rm -R ../FrACT ← this would give the iCloud services a hiccup, so mv to trash
pathFractTrash=$HOME/Library/Mobile\ Documents/com\~apple\~CloudDocs/.Trash/FrACT
rm -R $pathFractTrash # need to delete first, otherwise mv doesn't work
mv -fv ../FrACT $pathFractTrash

jake release

rm -R Build/Release/FrACT/Frameworks # we don't need this
rm -R Build/Release/FrACT/CommonJS.environment # nor this
mv Build/Release/FrACT ../ # move it up, creating the "capp" folder
rm -R Build # and get rid of the rest of the built items

cp webAppServiceWorker.js ../FrACT/

osascript -e 'display notification "Done."'
